src/core/thread/child.cpp

```diff
@@ -200,65 +200,41 @@ Error Child::RemoveIp6Address(const Ip6::Address &aAddress)
         ExitNow();
     }
 
-    for (index = 0; index < kNumIp6Addresses; index++)
-    {
-        VerifyOrExit(!mIp6Address[index].IsUnspecified());
-
-        if (mIp6Address[index] == aAddress)
-        {
-            error = kErrorNone;
-            break;
-        }
-    }
-
-    SuccessOrExit(error);
-
-    for (; index < kNumIp6Addresses - 1; index++)
-    {
-        mIp6Address[index] = mIp6Address[index + 1];
-    }
+    entry = mIp6Addresses.Find(aAddress);
+    VerifyOrExit(entry != nullptr);
 
-    mIp6Address[kNumIp6Addresses - 1].Clear();
+    mIp6Addresses.Remove(*entry);
+    error = kErrorNone;
```

Comment:
I think there's an existing issue here regarding how the MLR state of address entries is tracked.  It uses the bit vectors `mMlrToRegisterMask` and `mMlrRegisteredMask`. These masks track MLR states of each entry based on on its index within the array.

The `RemoveIp6Address()` method can shift existing entries within the list, altering their indexes. This can invalidate the MLR state bit vectors. I'll address this in a separate PR.